File: 19. Remove Nth Node From End of List/RemoveNthFromEnd.java
public class RemoveNthFromEnd {

    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        int sz = 0;
        ListNode curr = head;
        while (curr != null){
            curr = curr.next;
            sz++;
        }
        
        if (sz < 2)
            return null;
        
        if (sz == n){
            head = head.next;
            return head;
        }

        curr = head;
        for (int i = 1; i < sz - n; i++)
            curr = curr.next;
        
        ListNode nextToCurr = curr.next;
        curr.next = nextToCurr.next;
        nextToCurr.next = null;

        return head;
    }  
}

File: 19. Remove Nth Node From End of List/RemoveNthFromEnd.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    
    # O(n) time, one pass
    # O(1) space,
    # Approach: two pointers, 
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        first, second = dummy, dummy
        
        while n > 0:
            first = first.next
            n -= 1
        
        while first.next:
            first = first.next
            second = second.next
            
        prev = second.next
        second.next = prev.next
        prev.next = None
        
        return dummy.next
            
        